### PR TITLE
Completion, definition, highlight support for attributes which have references with settings

### LINKED ISSUE #1427: Completion, definition, highlight support for attributes which have references with settings
I noticed that there are a lof XML which declare some references between attribute like :

 * docbook with `xref@linkend` -> `*/@id` :

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE book PUBLIC "-//OASIS//DTD DocBook XML V4.4//EN" "http://www.docbook.org/xml/4.4/docbookx.dtd">
<book>
    <chapter id="chapter-1">

        <xref linkend="chapter-1" />

    </chapter>

    <chapter id="chapter-2">

    </chapter>
</book>
```

 * TEI with `corresp` and `resp` attributes `@corresp` -> `@xml:id`:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://www.tei-c.org/release/xml/tei/custom/schema/relaxng/tei_lite.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<TEI xmlns="http://www.tei-c.org/ns/1.0">
  <teiHeader>  
    <fileDesc>
      <titleStmt>
        <title>Title</title>
      </titleStmt>
      <publicationStmt>
        <p>Publication information</p>  
      </publicationStmt>
      <sourceDesc>
        <p>Information about the source</p>
      </sourceDesc>
    </fileDesc>
  </teiHeader>
  <text>
    <body xml:id="body-id">
      <p xml:id="p-id" >Some text here.</p>
      <anchor corresp="#body-id"></anchor>
    </body>
  </text>
</TEI>
```

Docbook uses ID, IDREF for xlink, so we could use this info (on the future), but this issue is to declare those links with settings

 * for docbook, we could declare link like this:

```json
  "xml.references": [
    {
      "pattern": "**/*.xml",
      "expressions": [
        {
          "from": "xref/@linkend",
          "to": "@id"
        }
      ]
    }
  ]
```
 * for TEI, we could declare link like this:

```json
"xml.references": [
    {
      "pattern": "**/*.xml",
      "expressions": [
        {
          "prefix": "#",
          "from": "@resp",
          "to": "persName/@xml:id"
        },
        {
          "prefix": "#",
          "from": "@corresp",
          "to": "@xml:id"
        }
      ]
    }
  ]
```

 * for Pretext, we could declare link like this:

```json
"xml.references": [
		{
			"pattern": "**/*.ptx",
			"expressions": [
				{
					"from": "xref/@ref",
					"to": "@xml:id"
				}
			]
		}
	]
```

This settings provides the capability to declare references betwwen attributes and will support (without coding something):

 * completion for referenced attributes
 * definition for a referenced attribute
 * highlight to hoghlight the refrenced attribute

### PR DESCRIPTION
**Pull request by angelozerr**
Completion, definition, highlight support for attributes which have references with settings

Fixes #1427

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR requires https://github.com/redhat-developer/vscode-xml/pull/847

**Comment by angelozerr** [comment]
@JessicaJHee if you wish you can play with this PR, see my explanation https://github.com/eclipse/lemminx/issues/1427

For instance with this xml

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://www.tei-c.org/release/xml/tei/custom/schema/relaxng/tei_lite.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<TEI xmlns="http://www.tei-c.org/ns/1.0">
  <teiHeader>  
    <fileDesc>
      <titleStmt>
        <title>Title</title>
      </titleStmt>
      <publicationStmt>
        <p>Publication information</p>  
      </publicationStmt>
      <sourceDesc>
        <p>Information about the source</p>
      </sourceDesc>
    </fileDesc>
  </teiHeader>
  <text>
    <body xml:id="body-id">
      <p xml:id="p-id" >Some text here.</p>
      <anchor corresp="#body-id"></anchor>
    </body>
  </text>
</TEI>
```

and this settings:

```json
"xml.references": [
    {
      "pattern": "**/*.xml",
      "expressions": [
        {
          "prefix": "#",
          "from": "@resp",
          "to": "persName/@xml:id"
        },
        {
          "prefix": "#",
          "from": "@corresp",
          "to": "@xml:id"
        }
      ]
    }
  ]

```

you should see 

TEIReferencesDemo

**Comment by angelozerr** [comment]
@JessicaJHee please don't review my code, I need to clean my code and write tests. 

I would like just to have some feedback about this feature if it misses some configuration or another thing.

**Comment by JulieGiovacchini** [comment]
It is great ! Thank you so much, it will be so helpful for students !

**Comment by angelozerr** [comment]
> It is great ! Thank you so much, it will be so helpful for students !

Glad you like this feature. I need to polish my PR before merging it. Once it will be merged you could test it by installing prerelease https://github.com/redhat-developer/vscode-xml#contributing

It should be really nice to give us feedback before we release vscode-xml to improve,s fixes some bugs with this references feature.

**Review by JessicaJHee — COMMENTED** [review]
Tested on my end and works perfect! Just left a few suggestions.

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/references/XMLReferencesDefinitionExtensionsTest.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import org.eclipse.lsp4j.LocationLink;
+import org.junit.jupiter.api.Test;
+
+/**
+ * XML references definition tests.
+ *
+ * @author Angelo ZERR
+ */
+public class XMLReferencesDefinitionExtensionsTest extends AbstractCacheBasedTest {
+
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/references/XMLReferencesDefinitionExtensionsTest.java:38]: Good catch! Tests exists now.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/references/utils/XMLReferencesUtils.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+			String prefix = originAttrValue.substring(0, index);
+			if (!Objects.equal(prefix, targetNamespacePrefix)) {
+				return null;
+			}
+			return originAttrValue.substring(index + 1, originAttrValue.length());
+		}
+		return originAttrValue;
+	}
+
+	public static List<XMLReferenceExpression> findExpressionWichMatchesFrom(DOMAttr attr,
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/references/utils/XMLReferencesUtils.java]: fixed

**Comment by angelozerr** [comment]
Thanks @JessicaJHee for your great review!

**Pull request merged**